R/data.table.R

```diff
@@ -1522,7 +1522,11 @@ replace_dot_alias = function(e) {
         jvnames = sdvars
       }
     } else if (length(as.character(jsub[[1L]])) == 1L) {  # Else expect problems with <jsub[[1L]] == >
-      subopt = length(jsub) == 3L && (jsub[[1L]] == "[" || jsub[[1L]] == "[[") && (is.numeric(jsub[[3L]]) || jsub[[3L]] == ".N")
+      # g[[ only applies to atomic input, for now, was causing #4159
+      subopt = length(jsub) == 3L &&
+        (jsub[[1L]] == "[" ||
+           (jsub[[1L]] == "[[" && eval(call('is.atomic', jsub[[2L]]), envir = x))) &&
```

Comment:
why you need `eval(call("is.atomic", jsub[[2L]]), envir=x)` here and not just `eval(is.atomic(jsub[[2L]]), envir=x)`?

Reply:
I didn't expect that would work. Much nicer.

Reply:
Oh, I spoke too soon. Turning it off for the other instance failed the test.

Reply:
I think this will run `is.atomic` on the `name` object `q[[2L]]` instead of its value

Reply:
ah, right